common/celo_types.go

```diff
@@ -9,9 +9,31 @@ var (
 )
 
 type ExchangeRates = map[Address]*big.Rat
+type IntrinsicGasCosts = map[Address]uint64
 
 type FeeCurrencyContext struct {
-	ExchangeRates ExchangeRates
+	ExchangeRates     ExchangeRates
+	IntrinsicGasCosts IntrinsicGasCosts
+}
+
+func MaxAllowedIntrinsicGasCost(i IntrinsicGasCosts, feeCurrency *Address) (uint64, bool) {
+	intrinsicGas, ok := CurrencyIntrinsicGasCost(i, feeCurrency)
+	if !ok {
+		return 0, false
+	}
+	// allow the contract to overshoot 2 times the deducted intrinsic gas
+	// during execution
+	return intrinsicGas * 3, true
+}
+func CurrencyIntrinsicGasCost(i IntrinsicGasCosts, feeCurrency *Address) (uint64, bool) {
+	if feeCurrency == nil {
+		return 0, true
+	}
```

Comment:
I had to look up why `feeCurrency` cannot be `nil` here. I think a comment would be helpful. Or maybe we can even panic.

Reply:
The reason it is there is to guard the pointer deref just below and because I thought it would make sense that the additional intrinsic gas cost of a `nil` currency is 0.
This also translates to an additional fee-currency specific max gas allowance for EVM calls as used by the `MaxAllowedIntrinsicGasCost` for the credit/debit allowance.
This then will be 0 and will cause the EVM calls to revert with out of gas, if ever called with a `nil` currency.

However this is all theoretical and currently there is no code-path where it comes to this, since the `CreditFees` and `DebitFees` are only called in non-nil fee-currency code-branches.

I will put the above in a comment.